Police-Data-Accessibility-Project/data-sources-app-v2
# Fix bug, create DynamicRequestDTO, fix typo in "DynamicMagicMock"

#### Fixes

* https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/394

#### Description

* Fixes bug causing record types to not properly narrow search as expected.
* Updated documentation
* Additional improvements as described in Leave It Better Than You Found It section.

#### Testing

* Run tests in `/tests` directory, confirm all function as expected.
* Run `/search` endpoint with record categories specified -- confirm these narrow down the list of results as expected.

#### Performance

* Use of dynamic logic adds a small but negligible amount of overhead. 

#### Docs

* Not applicable.

#### Breaking Changes

* None.

#### Leave It Better Than You Found It

The reason this occurred was because of a change in the design to accommodate the requirements of allowing multiple record categories. Thus, "record\_category" (singular) was changed to become "record\_categories" (plural). This change was reflected in some of the functions and the categories, but it was not reflected in either the request get arguments or in my tests (which still passed the value as "record\_category" despite the nomenclature changing).

To accommodate this and reduce the number of places I had to repeat the nomenclature, I created a Dynamic Requests Data Transfer Object (DTO), which would pull requests from a query based exactly on the name of the attributes the object was provided. Thus, if the attribute changes, what it pulls would change as well. The rationale was that this would make it less likely for a mistake such as that to go unnoticed.

This has an additional benefit of simplifying Resource logic as well as middleware and middleware tests, as the inclusion of a DTO which encapsulates multiple arguments means fewer attributes need to be passed from location to location; they are all contained within the DTO.
